File: julia_simulation/aggregate.jl
# Aggregate bootstrap results and compute confidence intervals.
#
# Usage: julia aggregate.jl <input_dir>
#
# Arguments:
#   input_dir - Directory containing res_*.csv files from bootstrap.jl
#
# Output: results.csv in the current working directory

using DelimitedFiles
using Statistics

input_dir = ARGS[1]

println("Aggregating bootstrap results")
println("  Input directory: $input_dir")

# Find all result files
files = filter(f -> startswith(f, "res_") && endswith(f, ".csv"), readdir(input_dir))
sort!(files)
println("  Found $(length(files)) result files")

if isempty(files)
    error("No result files found!")
end

# Read all results — extract estimation errors and first 3 coefficients
errors = Float64[]
betas_1 = Float64[]
betas_2 = Float64[]
betas_3 = Float64[]

for file in files
    path = joinpath(input_dir, file)
    lines = readlines(path)
    if length(lines) < 2
        continue
    end
    values = split(lines[2], ",")
    push!(errors, parse(Float64, values[2]))   # estimation_error
    push!(betas_1, parse(Float64, values[3]))   # beta_1
    push!(betas_2, parse(Float64, values[4]))   # beta_2
    push!(betas_3, parse(Float64, values[5]))   # beta_3
end

n_reps = length(errors)

# Compute bootstrap confidence intervals (percentile method)
function ci_95(x)
    sorted = sort(x)
    lo = sorted[max(1, Int(floor(0.025 * length(sorted))))]
    hi = sorted[min(length(sorted), Int(ceil(0.975 * length(sorted))))]
    return (lo, hi)
end

# Write summary
open("results.csv", "w") do io
    println(io, "variable,mean,std,ci_lower,ci_upper")
    for (name, vals) in [("beta_1", betas_1), ("beta_2", betas_2), ("beta_3", betas_3)]
        lo, hi = ci_95(vals)
        println(io, "$name,$(mean(vals)),$(std(vals)),$lo,$hi")
    end
    println(io, "estimation_error,$(mean(errors)),$(std(errors)),,")
end

println("\nBootstrap Results ($(n_reps) replications):")
println("  β₁: $(round(mean(betas_1), digits=4)) [$(round(ci_95(betas_1)[1], digits=4)), $(round(ci_95(betas_1)[2], digits=4))]")
println("  β₂: $(round(mean(betas_2), digits=4)) [$(round(ci_95(betas_2)[1], digits=4)), $(round(ci_95(betas_2)[2], digits=4))]")
println("  β₃: $(round(mean(betas_3), digits=4)) [$(round(ci_95(betas_3)[1], digits=4)), $(round(ci_95(betas_3)[2], digits=4))]")
println("  Mean estimation error: $(round(mean(errors), digits=6))")
println("\nResults saved to: results.csv")

File: python_simulation/aggregate.py
#!/usr/bin/env python3
"""
Aggregate Monte Carlo pricing results from individual simulations.

Usage: python aggregate.py <input_dir>

Arguments:
  input_dir - Directory containing res_*.csv files from price_option.py

Output: results.csv in the current working directory
"""

import csv
import glob
import os
import sys


def main():
    if len(sys.argv) != 2:
        print("Usage: python aggregate.py <input_dir>", file=sys.stderr)
        sys.exit(1)

    input_dir = sys.argv[1]
    print("Aggregating pricing results")
    print(f"  Input directory: {input_dir}")

    files = sorted(glob.glob(os.path.join(input_dir, "res_*.csv")))
    print(f"  Found {len(files)} result files")

    if not files:
        print("No result files found!", file=sys.stderr)
        sys.exit(1)

    # Read all results
    prices = []
    ses = []
    for path in files:
        with open(path, newline="") as f:
            reader = csv.DictReader(f)
            for row in reader:
                prices.append(float(row["price"]))
                ses.append(float(row["se"]))

    n = len(prices)
    mean_price = sum(prices) / n
    # SE of the combined estimate (average of independent estimates)
    combined_se = (sum(s**2 for s in ses) / n**2) ** 0.5
    min_price = min(prices)
    max_price = max(prices)

    # Write summary
    with open("results.csv", "w", newline="") as f:
        writer = csv.writer(f)
        writer.writerow(["n_simulations", "mean_price", "combined_se", "min_price", "max_price"])
        writer.writerow([n, f"{mean_price:.6f}", f"{combined_se:.6f}", f"{min_price:.6f}", f"{max_price:.6f}"])

    print(f"\nAggregated Results:")
    print(f"  Simulations:  {n}")
    print(f"  Mean price:   {mean_price:.4f}")
    print(f"  Combined SE:  {combined_se:.4f}")
    print(f"  Range:        [{min_price:.4f}, {max_price:.4f}]")
    print(f"\nResults saved to: results.csv")


if __name__ == "__main__":
    main()
